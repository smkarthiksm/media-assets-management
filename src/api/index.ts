import { sampleJwtToken, setJwtToken } from '../components/utilities/utility';
import {
  AllFile,
  AudioFile,
  VideoFile,
  mockAudioFiles,
  mockVideoFiles,
} from '../stubbs/files';
import { loginCreds } from '../stubbs/login-stub';

export const getSession = (e: string, p: string): Promise<boolean> => {
  const isUserExist = loginCreds.some(
    ({ email, password }) => email === e && password === p,
  );

  return new Promise((res, rej) => {
    setTimeout(() => {
      if (isUserExist) {
        setJwtToken(sampleJwtToken);
        return res(isUserExist);
      }
      return rej(new Error('Invalid creds'));
    }, 4000);
  });
};

export const getAllFiles = (): Promise<AllFile[]> => {
  return new Promise((res) => {
    setTimeout(() => {
      return res([...mockAudioFiles, ...mockVideoFiles]);
    }, 100);
  });
};

export const getAudioFiles = (): Promise<AudioFile[]> => {
  return new Promise((res) => {
    setTimeout(() => {
      return res(mockAudioFiles);
    }, 4000);
  });
};

export const getVideoFiles = (): Promise<VideoFile[]> => {
  return new Promise((res) => {
    setTimeout(() => {
      return res(mockVideoFiles);
    }, 4000);
  });
};
