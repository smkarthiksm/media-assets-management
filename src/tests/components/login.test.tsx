import '@testing-library/jest-dom';
import { renderWithProviders } from '../test-utility';
import LoginComponent from '../../components/login/login';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import * as Api from '../../api';

describe('LoginComponent', () => {
  test('should render', () => {
    const { container } = renderWithProviders(<LoginComponent />);
    expect(container).toMatchSnapshot();
  });

  test('should update text fields on input change', () => {
    const { store } = renderWithProviders(<LoginComponent />);

    const emailInput = screen.getByPlaceholderText('Email');

    expect(store.getState().auth.email).toEqual('');
    fireEvent.change(emailInput, {
      target: { value: 'simon.nixon@gmail.com' },
    });
    expect(store.getState().auth.email).toEqual('simon.nixon@gmail.com');
  });

  test('should disable the login button when email is invalid or password is empty', () => {
    renderWithProviders(<LoginComponent />);

    const emailInput = screen.getByPlaceholderText('Email');
    const passwordInput = screen.getByPlaceholderText('Password');

    let submitButton = screen.getByRole('button');
    fireEvent.change(emailInput, {
      target: { value: '' },
    });
    fireEvent.change(passwordInput, {
      target: { value: '' },
    });
    expect(submitButton).toBeDisabled();

    submitButton = screen.getByRole('button');
    fireEvent.change(emailInput, {
      target: { value: 'invalid' },
    });
    fireEvent.change(passwordInput, {
      target: { value: '' },
    });
    expect(submitButton).toBeDisabled();

    submitButton = screen.getByRole('button');
    fireEvent.change(emailInput, {
      target: { value: 'simon.nixon@gmail.com' },
    });
    fireEvent.change(passwordInput, {
      target: { value: '' },
    });
    expect(submitButton).toBeDisabled();

    submitButton = screen.getByRole('button');
    fireEvent.change(emailInput, {
      target: { value: 'simon.nixon@gmail.com' },
    });
    fireEvent.change(passwordInput, {
      target: { value: 'dummyPassword' },
    });
    expect(submitButton).toBeEnabled();
  });

  test('should show error message when login fails', async () => {
    const { store } = renderWithProviders(<LoginComponent />);

    const emailInput = screen.getByPlaceholderText('Email');
    const passwordInput = screen.getByPlaceholderText('Password');

    fireEvent.change(emailInput, {
      target: { value: 'dummyemail@gmail.com' },
    });
    fireEvent.change(passwordInput, {
      target: { value: 'dummyPassword' },
    });

    const mockFunction = jest
      .spyOn(Api, 'getSession')
      .mockReturnValue(new Promise((res, rej) => rej(false)));

    const submitButton = screen.getByRole('button');
    await waitFor(() => fireEvent.click(submitButton));

    expect(store.getState().auth.loginError).toBeTruthy();
    expect(screen.getByText('Account not found!')).toBeInTheDocument();
    mockFunction.mockRestore();
  });

  test('should show error message when login fails', async () => {
    const { store } = renderWithProviders(<LoginComponent />);

    const emailInput = screen.getByPlaceholderText('Email');
    const passwordInput = screen.getByPlaceholderText('Password');

    fireEvent.change(emailInput, {
      target: { value: 'dummyemail@gmail.com' },
    });
    fireEvent.change(passwordInput, {
      target: { value: 'dummyPassword' },
    });

    const mockFunction = jest
      .spyOn(Api, 'getSession')
      .mockReturnValue(new Promise((res) => res(true)));

    const submitButton = screen.getByRole('button');
    await waitFor(() => fireEvent.click(submitButton));

    expect(store.getState().auth.loginError).toBeFalsy();
    mockFunction.mockRestore();
  });
});
