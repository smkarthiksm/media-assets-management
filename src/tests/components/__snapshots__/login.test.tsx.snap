// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should render 1`] = `
<div>
  <div
    class="login-container"
  >
    <div
      class="logo-container"
    >
      <div
        class="logo"
      />
      <div
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        Media Assets Management
      </div>
      <div
        class="seperator mt-2"
      />
    </div>
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root input-field mt-4 css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root input-field mt-3 css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-ihdtdm"
            >
              <span
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth mt-3 css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Login
      </button>
    </div>
  </div>
</div>
`;
